### PR TITLE
Log prometheus metric registration error and fix CRD metric names

### PR DESCRIPTION
**Pull request by roycaihw**
/kind bug

When we call `prometheus.Register(c Collector)`, we didn't handle potential error for the registration. Some metrics failed to get registered due to [invalid name](https://github.com/kubernetes/kubernetes/blob/895f483fdfba055573681ef067e16d60df7985f8/vendor/github.com/prometheus/common/model/metric.go#L93) / [duplicate](https://github.com/kubernetes/kubernetes/blob/895f483fdfba055573681ef067e16d60df7985f8/vendor/github.com/prometheus/client_golang/prometheus/registry.go#L290) metrics: [gist of errors in a local cluster](https://gist.github.com/roycaihw/1bbb77d38e198b2528651b660da6c9da). As a result the metrics don't get exposed on `/metrics` endpoint.

This PR adds log for the errors and fixes some of the invalid CRD metric names. The name change should be backwards compatible since the metrics were never exposed.

**Special notes for your reviewer**:
Alternatively we could panic on the errors. However, there are errors caused by duplication (see the [gist](https://gist.github.com/roycaihw/1bbb77d38e198b2528651b660da6c9da) above) that I don't fully understand. Also there are [metrics](https://gist.github.com/roycaihw/57c95d70e00a169223e2f73cde2fee6d) with invalid name not triggering error in my local cluster. I didn't fix those metrics. 

**Does this PR introduce a user-facing change?**:

```release-note
Prometheus metrics for crd_autoregister, crd_finalizer and crd_naming_condition_controller are exported.
```

/sig api-machinery
/cc @logicalhan @mbohlool @sttts

### CODE REVIEW

**Review by logicalhan — APPROVED** [review]
This looks great, I really appreciate you digging into this. I like the idea about surfacing the registration errors, it'll help us go back and fix all these things. 
/lgtm

File: pkg/master/controller/crdregistration/crdregistration_controller.go [hunk 1]
```diff
@@ -68,7 +68,7 @@ func NewAutoRegistrationController(crdinformer crdinformers.CustomResourceDefini
 		crdSynced:              crdinformer.Informer().HasSynced,
 		apiServiceRegistration: apiServiceRegistration,
 		syncedInitialSet:       make(chan struct{}),
-		queue:                  workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "crd-autoregister"),
+		queue:                  workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "crd_autoregister"),
```
Comment (logicalhan) [inline pkg/master/controller/crdregistration/crdregistration_controller.go]: nit: can we make this `crd_autoregistration_controller` for consistency?

**Review by logicalhan — APPROVED** [review]

File: pkg/master/controller/crdregistration/crdregistration_controller.go (hunk 1 above)
Comment (roycaihw) [reply pkg/master/controller/crdregistration/crdregistration_controller.go]: Done

**Comment by roycaihw** [comment]
/retest

/assign @lavalamp 
for `pkg/OWNERS` approval

**Comment by lavalamp** [comment]
Doesn't this break anyone using the existing metrics? I don't know what we promise but 0 releases of deprecation doesn't seem right?

@deads2k or @liggitt, opinions?

**Comment by liggitt** [comment]
> Doesn't this break anyone using the existing metrics? I don't know what we promise but 0 releases of deprecation doesn't seem right?

I thought the invalid names failed to register and didn't get surfaced as metrics, meaning there can be no consumers to break

**Comment by roycaihw** [comment]
> I thought the invalid names failed to register and didn't get surfaced as metrics, meaning there can be no consumers to break

Yes. Prometheus only [registers a metric if there is no error](https://github.com/kubernetes/kubernetes/blob/611c3510d82941bb56c1d3c13d99521c70846260/vendor/github.com/prometheus/client_golang/prometheus/registry.go#L301). Those problematic metrics didn't exist in `/metrics` endpoint

**Comment by logicalhan** [comment]
They never worked, so no one can have depended on them.

**Comment by lavalamp** [comment]
Thanks for the added context, I think it wasn't clear when I first saw this that the names hadn't be exported previously :)

(no need to block on me next time I'm OOO for a long time, by the way)

/approve
/lgtm

File: pkg/master/controller/crdregistration/crdregistration_controller.go [hunk 2]
```diff
@@ -68,7 +68,7 @@ func NewAutoRegistrationController(crdinformer crdinformers.CustomResourceDefini
 		crdSynced:              crdinformer.Informer().HasSynced,
 		apiServiceRegistration: apiServiceRegistration,
 		syncedInitialSet:       make(chan struct{}),
-		queue:                  workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "crd-autoregister"),
+		queue:                  workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "crd_autoregistration_controller"),
```
Comment (sttts) [inline pkg/master/controller/crdregistration/crdregistration_controller.go]: to understand: dash was never allowed and this metric was never registered?

Comment (sttts) [reply pkg/master/controller/crdregistration/crdregistration_controller.go]: nvmd. It's the queue name.

File: pkg/util/workqueue/prometheus/prometheus.go [hunk 3]
```diff
@@ -37,7 +38,9 @@ func (prometheusMetricsProvider) NewDepthMetric(name string) workqueue.GaugeMetr
 		Name:      "depth",
 		Help:      "Current depth of workqueue: " + name,
 	})
-	prometheus.Register(depth)
+	if err := prometheus.Register(depth); err != nil {
+		klog.Errorf("failed to register depth metric %v: %v", name, err)
```
Comment (sttts) [inline pkg/util/workqueue/prometheus/prometheus.go]: shouldn't this me a more fatal error actually? It's a programming error. Those names are static strings, i.e. they don't come from the user.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/controller/finalizer/crd_finalizer.go [hunk 4]
```diff
@@ -79,7 +79,7 @@ func NewCRDFinalizer(
 		crdLister:      crdInformer.Lister(),
 		crdSynced:      crdInformer.Informer().HasSynced,
 		crClientGetter: crClientGetter,
-		queue:          workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "CustomResourceDefinition-CRDFinalizer"),
+		queue:          workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "crd_finalizer"),
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/controller/finalizer/crd_finalizer.go]: crd_finalizer_controller?

File: pkg/util/workqueue/prometheus/prometheus.go (hunk 3 above)
Comment (roycaihw) [reply pkg/util/workqueue/prometheus/prometheus.go]: there are more failures caused by duplication that I didn't dig into (see the [gist](https://gist.github.com/roycaihw/1bbb77d38e198b2528651b660da6c9da)). We can have a followup issue to clean those up

Comment (roycaihw) [reply pkg/util/workqueue/prometheus/prometheus.go]: it's not a complete list. kcm/ccm may also have such failure

File: staging/src/k8s.io/apiextensions-apiserver/pkg/controller/finalizer/crd_finalizer.go (hunk 4 above)
Comment (roycaihw) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/controller/finalizer/crd_finalizer.go]: crd_finalizer_controller sounds like a controller for crd finalizer to me

also we tend to omit "controller" from the name for other in-tree controllers: https://github.com/kubernetes/kubernetes/search?q=NewNamedRateLimitingQueue&unscoped_q=NewNamedRateLimitingQueue

File: pkg/util/workqueue/prometheus/prometheus.go (hunk 3 above)
Comment (roycaihw) [reply pkg/util/workqueue/prometheus/prometheus.go]: created #73551

**Pull request merged**